egs/wsj/s5/steps/nnet3/components.py

```diff
@@ -213,14 +214,59 @@ def AddSigmoidLayer(config_lines, name, input, self_repair_scale = None):
     return {'descriptor':  '{0}_sigmoid'.format(name),
             'dimension': input['dimension']}
 
+def AddUnfoldedRnnLayer(config_lines, name, input, output_dim, num_unfolded_times, ng_affine_options = "", norm_target_rms = 1.0, rnn_delay = -1, self_repair_scale = None):
+    components = config_lines['components']
+    component_nodes = config_lines['component-nodes']
+
+    components.append("component name={0}_affine type=NaturalGradientAffineComponent input-dim={1} output-dim={2} {3}".format(name, input['dimension'] + output_dim, output_dim, ng_affine_options))
+    self_repair_string = "self-repair-scale={0:.10f}".format(self_repair_scale) if self_repair_scale is not None else ''
+    components.append("component name={0}_relu type=RectifiedLinearComponent dim={1}".format(name, output_dim, self_repair_string))
+    components.append("component name={0}_renorm type=NormalizeComponent dim={1} target-rms={2}".format(name, output_dim, norm_target_rms))
+
+    for i in range(1, num_unfolded_times + 1):
+        # a node with its "x" index replaced with a value other than 0 (e.g., -1) is NOT computable                                                                                       
+        # in this network, so the recurrent input to "<name>_renorm<num_unfolded_times>_t" would be zero
+        recurrent_descriptor_string = "Offset({0}_renorm{1}_t, {2})".format(name, i + 1, rnn_delay) if i < num_unfolded_times else "ReplaceIndex({0}_renorm{1}_t, x, -1)".format(name, i)
```

Comment:
Please do not use `ReplaceIndex({0}_renorm{1}_t, x, -1)` just to get a zero. A very simple way to avoid this complication would be to just separate the W_x and W_y matrices. All the multiplications of W_x *[x(t), x(t+3) x(t+6)] can be done simultaneously so you will still be able to parallelize well.